### PR TITLE
Turn BidirectionalLM into a more-general LanguageModel class

### PR DESCRIPTION
**Pull request by nelson-liu**
Fixes https://github.com/allenai/allennlp/issues/2255

This PR replaces the `BidirectionalLM` class with a more-general `LanguageModel` that can be used in either the unidirectional/forward setting or the bidirectional setting.

It also accordingly replaces the `BidirectionalLanguageModelTokenEmbedder` with a `LanguageModelTokenEmbedder`.

Also fixes bug in the experiment_unsampled.jsonnet config that was preventing a test from actually being unsampled.

TODO:

- [x] test the unidirectional case
- [x] properly deprecate `BidirectionalLM` and `BidirectionalLanguageModelTokenEmbedder` 
- [x] check docs for accuracy
- [x] fix user-facing training configs

### CODE REVIEW

**Comment by nelson-liu** [comment]
I don't think i'll have this finished for the near future's minor release. Thus, it'd be great to merge #2253 before this and the next release.

**Comment by brendan-ai2** [comment]
I like this idea. My main concern would be minimizing disruption, e.g. keeping trained models working. Please ping me when you're ready for a review!

**Comment by nelson-liu** [comment]
ok @brendan-ai2 this should be ready to look at! No rush, though. I'm also planning on taking this code and actually running it on some data to ensure that the perplexities look more-or-less correct.

one question (for anyone): if we deprecate a class, should we keep its tests around? or is that unnecessary?

**Comment by matt-gardner** [comment]
@nelson-liu, on your deprecation question.  @schmmd and I talked about this a bit a while ago, and we think it'd be good to eventually make allennlp-internal warnings into test failures.  So, any test that emits a deprecation warning either has to catch/suppress that warning or fail.  If you're moving the functionality of a class to a new class, you should also move any applicable tests to the new class.  Whether you keep around a test on the deprecated class, that probably only checks that it emits a deprecation warning, is up to you.

**Comment by matt-peters** [comment]
Probably should have looked at this earlier but was out over the holidays so apologies for the late feedback.  What is the reasoning for explicitly adding `ShuffledSentence` to the names?  There is nothing inside the BidirectionalLM class that restricts it to shuffled sentences vs longer contexts, and I have used the calypso version to train on both long contexts and the shuffled Billion Word Benchmark.  

The assumptions about shuffling require changes in the data preparation / iterator and statefulness vs non-statefulness of the contextual encoder.  So it seems to me the abstractions should be `LanguageModel` (with bidirectional option), and adding a `StatefullIterator` that will read very long contexts, chunk them into smaller pieces and do the alignment from batch to batch?

**Comment by nelson-liu** [comment]
That's a good point @matt-peters , will edit.

**Comment by nelson-liu** [comment]
i think my thought process at the time was that rewrapping the hidden states with new tensors to detach them from their history would have to be handled in the `LanguageModel` class, but that's a moot point with the stateful contextualizers in allennlp (was thinking of the pytorch word level lm example and the batching done there...) Thanks for catching that!

**Comment by brendan-ai2** [comment]
I would suggest that we keep `BidirectionalLanguageModel` as a shorthand for setting `bidirectional=True` indefinitely. Creating churn for users by breaking APIs (which is effectively what changing a public name is) is rarely appreciated.

File: allennlp/models/language_model.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+from allennlp.modules.text_field_embedders import TextFieldEmbedder
+from allennlp.modules.sampled_softmax_loss import SampledSoftmaxLoss
+from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder
+from allennlp.nn.util import get_text_field_mask
+from allennlp.nn import InitializerApplicator
+
+
+class _SoftmaxLoss(torch.nn.Module):
+    """
+    Given some embeddings and some targets, applies a linear layer
```
Comment (matt-peters) [inline allennlp/models/language_model.py]: Can you explicitly document here in the constructor that this class does not add a padding ID, and that the input targets to forward should not include the padding id?

File: allennlp/models/bidirectional_lm.py [hunk 2]
```diff
@@ -94,211 +58,12 @@ def __init__(self,
                  num_samples: int = None,
                  sparse_embeddings: bool = False,
                  initializer: InitializerApplicator = None) -> None:
```
Comment (matt-peters) [inline allennlp/models/bidirectional_lm.py]: It is possible to not duplicate the input arguments from `LanguageModel` and just use `*args, **kwargs` so changes in `LanguageModel` are automatically reflected here?  Or does that break `from_params`?

File: allennlp/models/language_model.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    vocab: ``Vocabulary``
+    text_field_embedder: ``TextFieldEmbedder``
+        Used to embed the indexed tokens we get in ``forward``.
+    contextualizer: ``Seq2SeqEncoder``
+        Used to "contextualize" the embeddings. As described above,
+        this encoder must not cheat by peeking ahead.
+    dropout: ``float``, optional (default: None)
+        If specified, dropout is applied to the contextualized embeddings before computation of
+        the softmax. The contextualized embeddings themselves are returned without dropout.
+    loss_scale: ``Union[float, str]``, optional (default: 1.0)
```
Comment (matt-peters) [inline allennlp/models/language_model.py]: We can remove this argument, I don't think it's useful.

File: allennlp/models/language_model.py (hunk 1 above)
Comment (nelson-liu) [reply allennlp/models/language_model.py]: mind if i do this in another pr? would rather that this changeset not be larger than it has to for the core changes for configurable bidirectionality.

File: allennlp/models/language_model.py (hunk 3 above)
Comment (nelson-liu) [reply allennlp/models/language_model.py]: same as above

File: allennlp/models/bidirectional_lm.py (hunk 2 above)
Comment (nelson-liu) [reply allennlp/models/bidirectional_lm.py]: that would be great if it worked, i'll try it out. thanks for the suggestion

File: allennlp/models/language_model.py (hunk 1 above)
Comment (nelson-liu) [reply allennlp/models/language_model.py]: looks like it doesn't work :(

Comment (matt-peters) [reply allennlp/models/language_model.py]: Oh well.

**Comment by matt-peters** [comment]
This look good to me, although will let @brendan-ai2 take a look and approve as he ported over most of the code.

File: allennlp/models/language_model.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                        'forward_loss': forward_loss * scale_factor / num_targets.float(),
+                        'backward_loss': (backward_loss * scale_factor / num_targets.float()
+                                          if backward_loss is not None else None)
+                })
+            else:
+                # average_loss zero tensor, return it for all
+                return_dict.update({
+                        'loss': average_loss,
+                        'forward_loss': average_loss,
+                        'backward_loss': average_loss
```
Comment (brendan-ai2) [inline allennlp/models/language_model.py]: `None` if not bidirectional for consistency?

File: allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py [hunk 5]
```diff
@@ -1,26 +1,21 @@
...
-from allennlp.nn.util import remove_sentence_boundaries, get_text_field_mask, add_sentence_boundary_token_ids
-
-# Importing at runtime results in a circular dependency.
-if TYPE_CHECKING:
-    from allennlp.models.bidirectional_lm import BidirectionalLanguageModel
 
 
 @TokenEmbedder.register('bidirectional_lm_token_embedder')
 class BidirectionalLanguageModelTokenEmbedder(TokenEmbedder):
+    # pylint: disable=abstract-method
```
Comment (brendan-ai2) [inline allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py]: This shouldn't be needed. Looks like the problem is that we aren't extending the `LanguageModelTokenEmbedder` in the line above. This is good example of why keeping the tests is around is prudent. Fortunately you should be able to do this with minimal duplication. See https://github.com/allenai/allennlp/blob/511c846c9c779fdac0066dbae6765c1619037098/allennlp/tests/fixtures/bidirectional_lm/experiment.jsonnet for how to share config across jsonnet files if you aren't already familiar.

File: allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py [hunk 6]
```diff
@@ -1,26 +1,21 @@
...
     """
     Compute a single layer of representations from a bidirectional language model. This is done
     by computing a learned scalar average of the layers from the LM. Typically the LM's weights
     will be fixed, but they can be fine tuned by setting ``requires_grad``.
 
+    .. deprecated:: 0.8
+        ``BidirectionalLanguageModelTokenEmbedder`` was deprecated in version 0.8
+        and will be removed in version 0.10 .
```
Comment (brendan-ai2) [inline allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py]: As above.

File: allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet [hunk 7]
```diff
@@ -22,7 +22,6 @@
   },
   "model": {
     "type": "bidirectional_language_model",
-    "num_samples": 10,
```
Comment (brendan-ai2) [inline allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet]: Why did this change? I always get paranoid when tests change in a (mostly) refactoring PR... ;)

File: allennlp/tests/fixtures/language_model/experiment_unidirectional_unsampled.jsonnet [hunk 8]
```diff
@@ -0,0 +1,68 @@
+{
```
Comment (brendan-ai2) [inline allennlp/tests/fixtures/language_model/experiment_unidirectional_unsampled.jsonnet]: It would be great to reduce the duplication of this config with that in allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet and allennlp/tests/fixtures/language_model/experiment_unsampled.jsonnet.

File: allennlp/tests/models/language_model_test.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
+# pylint: disable=invalid-name,arguments-differ,abstract-method
+import numpy as np
+import pytest
+
+from allennlp.common.testing import ModelTestCase
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.params import Params
+from allennlp.models import Model
+
+class TestUnidirectionalLanguageModel(ModelTestCase):
```
Comment (brendan-ai2) [inline allennlp/tests/models/language_model_test.py]: Having an extra option for the directionality has made it difficult to see at a glance that the different tests are behaving as expected. I didn't do a great job with this when I added the transformer test, but in retrospect it should be pretty simple.

Roughly what needs to happen is a) set `expected_embedding_shape` in the constructor and b) check against this later. Similarly for `expected_backward_loss` and `breaking_contextualizer_direction`. Actually, I think the inherited test methods will be used too. (Worth verifying that works as expected.) So a lot of code can be removed.

As an aside, there is a balance to be struck between straightforward tests with modest duplication and parameterized ones. Here I'd lean towards parameterizing.

File: allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet (hunk 7 above)
Comment (nelson-liu) [reply allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet]: If the experiment is unsampled, isn't the right thing to do to specify a value of `None` to the `"num_samples"`? i.e., i'm pretty sure the `_SoftmaxLoss` wasn't being tested before?

`"num_samples": 10` is being added in `experiment.json` ( https://github.com/allenai/allennlp/blob/master/allennlp/tests/fixtures/bidirectional_lm/experiment.jsonnet#L5 ). Or did I overlook something?

File: allennlp/models/bidirectional_lm.py [hunk 10]
```diff
@@ -1,57 +1,20 @@
...
-                dim=-1
-        )
-
-        return torch.nn.functional.nll_loss(probs, targets.long(), reduction="sum")
-
-
 @Model.register('bidirectional-language-model',
                 deprecation_message=('The "bidirectional-language-model" name was '
                                      'deprecated in version 0.8 and will be removed'
                                      'in version 0.10 . '
```
[the remark this answers is not in the document]
Comment (nelson-liu) [reply allennlp/models/bidirectional_lm.py]: ok, so you don't want to deprecate `bidirectional-language-model`?

File: allennlp/models/language_model.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        # shape (batch_size, timesteps) masked with 0
+        if self._bidirectional:
+            forward_embeddings, backward_embeddings = lm_embeddings.chunk(2, -1)
+        else:
+            forward_embeddings = lm_embeddings
+            backward_embeddings = None
+        losses: List[torch.Tensor] = []
+        for idx, embedding, targets in ((0, forward_embeddings, forward_targets),
+                                        (1, backward_embeddings, backward_targets)):
+            if embedding is None and targets is None:
```
[the remark this answers is not in the document]
Comment (nelson-liu) [reply allennlp/models/language_model.py]: maybe i'm missing the point, but i don't think i could break the entire `for` statement out into a new function? or are you saying, break out the portion underneath the `if`? end result would still be an `if` in the `for`, though?

File: allennlp/models/bidirectional_lm.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (brendan-ai2) [reply allennlp/models/bidirectional_lm.py]: I don't hold an opinion on whether to apply the deprecation warning or not*. Encouraging standardizing on underscores is reasonable. However, I don't want to actually remove the name as it breaks the API. I think API breakage should be used only when strictly necessary.

* Depends on what we want to communicate as a team with deprecation warnings. I think it can be useful to communicate that "you shouldn't use this there is something better" without also saying "we're definitely going to remove this soon". This varies across projects.

[the remark this answers is not in the document]
Comment (nelson-liu) [reply allennlp/models/bidirectional_lm.py]: thanks for clarifying!

**Comment by nelson-liu** [comment]
@brendan-ai2 this is ready for another look, fyi

File: allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet (hunk 7 above)
Comment (brendan-ai2) [reply allennlp/tests/fixtures/bidirectional_lm/experiment_unsampled.jsonnet]: Good catch! I added the following to the description:

"Also fixes bug in the experiment_unsampled.jsonnet config that was preventing a test from actually being unsampled."

Even though it's a small fix code-wise, noting the change explicitly could end up being a useful breadcrumb later.

File: allennlp/tests/models/bidirectional_lm_test.py [hunk 12]
```diff
@@ -25,42 +27,17 @@ def test_forward_pass_runs_correctly(self):
...
-        result = self.model(**training_tensors)
-
-        assert set(result) == {"loss", "forward_loss", "backward_loss",
-                               "lm_embeddings", "noncontextual_token_embeddings", "mask"}
-
-        # The model should preserve the BOS / EOS tokens.
-        embeddings = result["lm_embeddings"]
-        # The BidirectionalLanguageModelTransformer uses input size * 2 as the output size unlike
-        # a bidirectional LSTM, which uses hidden size * 2.
-        assert tuple(embeddings.shape) == (2, 8, 32)
```
Comment (brendan-ai2) [inline allennlp/tests/models/bidirectional_lm_test.py]: Something seems off here. Shouldn't the expected dimension be set to `(2, 8, 32)`? Sorry if I'm missing something.

Comment (brendan-ai2) [reply allennlp/tests/models/bidirectional_lm_test.py]: It's concerning that the tests didn't fail...

Comment (nelson-liu) [reply allennlp/tests/models/bidirectional_lm_test.py]: So this experiment uses `experiment_bidirectional_unsampled.jsonnet`, which takes its config from `experiment_unsampled.jsonnet`. In that, we've got:

```
"contextualizer": {
        "type": "lstm",
        "bidirectional": true,
        "num_layers": 3,
        "input_size": 16,
        "hidden_size": 7,
    }
  },
```

So hidden size * 2 = 14?

Comment (nelson-liu) [reply allennlp/tests/models/bidirectional_lm_test.py]: the `TestBidirectionalLMTransformer` was moved to a different class, which is why it isn't there.

Comment (brendan-ai2) [reply allennlp/tests/models/bidirectional_lm_test.py]: D'oh. Sorry! Read that diff all wrong.

Comment (nelson-liu) [reply allennlp/tests/models/bidirectional_lm_test.py]: sorry it's so convoluted with stuff moving around everywhere, but also not completely disappearing...

File: training_config/bidirectional_language_model.jsonnet [hunk 13]
```diff
@@ -58,10 +58,11 @@ local BASE_ITERATOR = {
       //     "tokens": ["<S>", "</S>"],
       //     "token_characters": ["<>/S"]
       // },
```
Comment (brendan-ai2) [inline training_config/bidirectional_language_model.jsonnet]: Please update the path to this file in https://github.com/allenai/allennlp/blob/master/tutorials/how_to/training_transformer_elmo.md. You should also grep the codebase for any other references to this file. And recall that this won't help external users who may have references to this file already. Please consider this carefully when weighing whether a rename should be performed.

Comment (nelson-liu) [reply training_config/bidirectional_language_model.jsonnet]: done, thanks for pointing that out

**Comment by nelson-liu** [comment]
Thanks for your thorough comments, @brendan-ai2 ! Much appreciated.
